shipyard_utils/files.py

```diff
@@ -102,6 +104,107 @@ def determine_destination_full_path(
         destination_folder_name, destination_file_name)
     return destination_full_path
 
+# Functions for handling large files
+
+
+def compress_files(file_paths, destination_full_path, compression):
+    """
+    Given a list of files, compress all of them into a single file.
+    Keeps the existing directory structure in tact.
+    """
+    if f'.{compression}' in destination_full_path:
+        compressed_file_name = destination_full_path
+    else:
+        compressed_file_name = f'{destination_full_path}.{compression}'
+
+    if compression == 'zip':
+        with ZipFile(compressed_file_name, 'w') as zip:
+            for file in file_paths:
+                file = clean_folder_name(file.replace(os.getcwd(), ''))
+                zip.write(file)
+                print(
+                    f'Successfully compressed {file} into {compressed_file_name}')
+
+    if compression == 'tar.bz2':
+        with tarfile.open(compressed_file_name, 'w:bz2') as tar:
+            for file in file_paths:
+                file = clean_folder_name(file.replace(os.getcwd(), ''))
+                tar.add(file)
+                print(
+                    f'Successfully compressed {file} into {compressed_file_name}')
+
+    if compression == 'tar':
+        with tarfile.open(compressed_file_name, 'w') as tar:
+            for file in file_paths:
+                file = clean_folder_name(file.replace(os.getcwd(), ''))
+                tar.add(file)
+                print(
+                    f'Successfully compressed {file} into {compressed_file_name}')
+
+    if compression == 'tar.gz':
+        with tarfile.open(compressed_file_name, 'w:gz') as tar:
+            for file in file_paths:
+                file = clean_folder_name(file.replace(os.getcwd(), ''))
+                tar.add(file)
+                print(
+                    f'Successfully compressed {file} into {compressed_file_name}')
```

Comment:
Each of these can be their own function. Specifically the with clauses. 
It's unfortunate that the zip uses `write` and tar uses `add` for adding files, so those will have to be separate. 
But especially for tar, you can write a function to do everything in the with - all that code is the exact same, just the with changes. You can even pass the second `tarfile.open` parameter to the function, since that is the only change among all tar if clauses.

Reply:
@ericelsken To confirm, still keep the higher-level function of "compress_files" but make each sub-component it's own function?

Reply:
Yes. Essentially one for each compression algorithm. [](edited-with-Axolo)